Support operations on c10::complex and integer scalars

xuhdev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38422 Remove some unnecessary cast for complex numbers.
* **#38418 Support operations on c10::complex and integer scalars**

This is useful in reducing verbosity in c10::complex's general usage, and potentially also offers
performance benefits.

This brings back #34506 (which was made for std::complex).

Differential Revision: [D21587012](https://our.internmc.facebook.com/intern/diff/D21587012)

xuhdev committed:
Support operations on c10::complex and integer scalars

This is useful in reducing verbosity in c10::complex's general usage, and potentially also offers
performance benefits.

This brings back #34506 (which was made for std::complex).

[ghstack-poisoned]

xuhdev committed:
Update on "Support operations on c10::complex and integer scalars"


This is useful in reducing verbosity in c10::complex's general usage, and potentially also offers
performance benefits.

This brings back #34506 (which was made for std::complex).

[ghstack-poisoned]

zasdfgbnm commented on c10/test/util/complex_test_common.h:
```diff
@@ -389,6 +389,32 @@ MAYBE_GLOBAL void test_arithmetic() {
+template<typename T, typename int_t>
+static void test_binary_ops_for_int_type_(T real, T img, int_t num) {
+  c10::complex<T> c(real, img);
+  ASSERT_EQ(c + num, c10::complex<T>(real + num, img + num));
```
Shouldn't the result be `c10::complex<T>(real + num, img)`

xuhdev replied:
You are right, fixed.

xuhdev committed:
Update on "Support operations on c10::complex and integer scalars"


This is useful in reducing verbosity in c10::complex's general usage, and potentially also offers
performance benefits.

This brings back #34506 (which was made for std::complex).

[ghstack-poisoned]

zasdfgbnm commented on c10/util/complex_type.h:
```diff
@@ -404,6 +404,38 @@ constexpr bool operator!=(const T& lhs, const c10::complex<T>& rhs) {
 }
 
+
+// Define operators between integral scalars and c10::complex
```
Nit: Could you have a line commenting that this behavior is different from `std::complex` but we feel it useful?

xuhdev replied:
Added

zasdfgbnm commented on c10/util/complex_type.h:
```diff
@@ -404,6 +404,38 @@ constexpr bool operator!=(const T& lhs, const c10::complex<T>& rhs) {
+  typename std::enable_if_t<std::is_floating_point<fT>::value && std::is_integral<iT>::value, int> = 0
+
+template<typename fT, typename iT, COMPLEX_INTEGER_OP_TEMPLATE_CONDITION>
+c10::complex<fT> operator+(const c10::complex<fT>& a, const iT& b) { return a + static_cast<fT>(b); }
```
`constexpr`

xuhdev replied:
Added

zasdfgbnm commented on c10/test/util/complex_test_common.h:
```diff
@@ -389,6 +389,32 @@ MAYBE_GLOBAL void test_arithmetic() {
 }
 
+template<typename T, typename int_t>
+static void test_binary_ops_for_int_type_(T real, T img, int_t num) {
```
`C10_HOST_DEVICE `

xuhdev replied:
Added

xuhdev replied:
I had to remove this keyword, otherwise I keep getting error on Windows CI:

```
C:/Users/circleci/project\c10/test/util/complex_test_common.h(407): error: calling a __host__ function from a __host__ __device__ function is not allowed
          detected during instantiation of "void arithmetic::test_binary_ops_for_all_int_types_(T, T, int8_t) [with T=float]" 
(414): here

C:/Users/circleci/project\c10/test/util/complex_test_common.h(408): error: calling a __host__ function from a __host__ __device__ function is not allowed
          detected during instantiation of "void arithmetic::test_binary_ops_for_all_int_types_(T, T, int8_t) [with T=float]" 
(414): here

C:/Users/circleci/project\c10/test/util/complex_test_common.h(409): error: calling a __host__ function from a __host__ __device__ function is not allowed
          detected during instantiation of "void arithmetic::test_binary_ops_for_all_int_types_(T, T, int8_t) [with T=float]" 
(414): here

C:/Users/circleci/project\c10/test/util/complex_test_common.h(410): error: calling a __host__ function from a __host__ __device__ function is not allowed
          detected during instantiation of "void arithmetic::test_binary_ops_for_all_int_types_(T, T, int8_t) [with T=float]" 
(414): here
```

xuhdev committed:
Update on "Support operations on c10::complex and integer scalars"


This is useful in reducing verbosity in c10::complex's general usage, and potentially also offers
performance benefits.

This brings back #34506 (which was made for std::complex).

[ghstack-poisoned]

zasdfgbnm commented on c10/test/util/complex_test_common.h:
```diff
@@ -389,6 +389,32 @@ MAYBE_GLOBAL void test_arithmetic() {
+  ASSERT_EQ(c * num, c10::complex<T>(real * num, img * num));
+  ASSERT_EQ(num * c, c10::complex<T>(num * real, num * img));
+  ASSERT_EQ(c / num, c10::complex<T>(real / num, img / num));
+  ASSERT_EQ(num / c, c10::complex<T>(num / real, num / img));
```
I don't think this is correct, `a / z == a * conj(z) / |z|^2`

xuhdev replied:
Nice catch!

xuhdev committed:
Update on "Support operations on c10::complex and integer scalars"


This is useful in reducing verbosity in c10::complex's general usage, and potentially also offers
performance benefits.

This brings back #34506 (which was made for std::complex).

[ghstack-poisoned]

zasdfgbnm commented:
Thanks for improving this!

xuhdev committed:
Update on "Support operations on c10::complex and integer scalars"


This is useful in reducing verbosity in c10::complex's general usage, and potentially also offers
performance benefits.

This brings back #34506 (which was made for std::complex).

[ghstack-poisoned]

xuhdev committed:
Update on "Support operations on c10::complex and integer scalars"


This is useful in reducing verbosity in c10::complex's general usage, and potentially also offers
performance benefits.

This brings back #34506 (which was made for std::complex).

[ghstack-poisoned]

xuhdev committed:
Update on "Support operations on c10::complex and integer scalars"


This is useful in reducing verbosity in c10::complex's general usage, and potentially also offers
performance benefits.

This brings back #34506 (which was made for std::complex).

[ghstack-poisoned]

xuhdev committed:
Update on "Support operations on c10::complex and integer scalars"


This is useful in reducing verbosity in c10::complex's general usage, and potentially also offers
performance benefits.

This brings back #34506 (which was made for std::complex).

[ghstack-poisoned]

xuhdev committed:
Update on "Support operations on c10::complex and integer scalars"


This is useful in reducing verbosity in c10::complex's general usage, and potentially also offers
performance benefits.

This brings back #34506 (which was made for std::complex).

[ghstack-poisoned]

xuhdev committed:
Update on "Support operations on c10::complex and integer scalars"


This is useful in reducing verbosity in c10::complex's general usage, and potentially also offers
performance benefits.

This brings back #34506 (which was made for std::complex).

[ghstack-poisoned]

xuhdev committed:
Update on "Support operations on c10::complex and integer scalars"


This is useful in reducing verbosity in c10::complex's general usage, and potentially also offers
performance benefits.

This brings back #34506 (which was made for std::complex).

[ghstack-poisoned]

xuhdev commented on c10/test/util/complex_test_common.h:
```diff
@@ -389,6 +389,32 @@ MAYBE_GLOBAL void test_arithmetic() {
+  test_binary_ops_for_int_type_<T, int64_t>(real, img, i);
+}
+
+TEST(TestArithmeticIntScalar, All) {
```
@zasdfgbnm I ended up using the `TEST` macro to avoid the Windows CI error

zasdfgbnm replied:
That is OK, could you add a comment?

xuhdev replied:
Probably make a separate PR and add some Windows CUDA comments somewhere? I'm not sure what the true reason this is (but I'm aware that Windows CUDA has a lot of stern requirements). Perhaps @ezyang knows more about this? Thanks!

The Windows error is:

```
C:/Users/circleci/project\c10/test/util/complex_test_common.h(407): error: calling a __host__ function from a __host__ __device__ function is not allowed
          detected during instantiation of "void arithmetic::test_binary_ops_for_all_int_types_(T, T, int8_t) [with T=float]" 
(414): here

C:/Users/circleci/project\c10/test/util/complex_test_common.h(408): error: calling a __host__ function from a __host__ __device__ function is not allowed
          detected during instantiation of "void arithmetic::test_binary_ops_for_all_int_types_(T, T, int8_t) [with T=float]" 
(414): here

C:/Users/circleci/project\c10/test/util/complex_test_common.h(409): error: calling a __host__ function from a __host__ __device__ function is not allowed
          detected during instantiation of "void arithmetic::test_binary_ops_for_all_int_types_(T, T, int8_t) [with T=float]" 
(414): here

C:/Users/circleci/project\c10/test/util/complex_test_common.h(410): error: calling a __host__ function from a __host__ __device__ function is not allowed
          detected during instantiation of "void arithmetic::test_binary_ops_for_all_int_types_(T, T, int8_t) [with T=float]" 
(414): here
```

The change that makes the Windows error message go away:

```diff
diff --git a/c10/test/util/complex_test_common.h b/c10/test/util/complex_test_common.h
index 4fde8da2367e..fe61ca0e231d 100644
--- a/c10/test/util/complex_test_common.h
+++ b/c10/test/util/complex_test_common.h
@@ -390,27 +390,27 @@ MAYBE_GLOBAL void test_arithmetic() {
 }
 
 template<typename T, typename int_t>
-C10_HOST_DEVICE void test_binary_ops_for_int_type_(T real, T img, int_t num) {
+void test_binary_ops_for_int_type_(T real, T img, int_t num) {
   c10::complex<T> c(real, img);
   static_assert(c + num == c10::complex<T>(real + num, img), "");
   static_assert(num + c == c10::complex<T>(num + real, img), "");
@@ -403,7 +403,7 @@ inline void test_binary_ops_for_int_type_(T real, T img, int_t num) {
 } 

 template<typename T>
-C10_HOST_DEVICE void test_binary_ops_for_all_int_types_(T real, T img, int8_t i) {
+void test_binary_ops_for_all_int_types_(T real, T img, int8_t i) {
   test_binary_ops_for_int_type_<T, int8_t>(real, img, i);
   test_binary_ops_for_int_type_<T, int16_t>(real, img, i);
   test_binary_ops_for_int_type_<T, int32_t>(real, img, i);
   test_binary_ops_for_int_type_<T, int64_t>(real, img, i);
 }
 
-MAYBE_GLOBAL void test_arithmetic_int_scalar() {
+TEST(TestArithmeticIntScalar, All) {
   test_binary_ops_for_all_int_types_<float>(1.0, 0.1, 1);
   test_binary_ops_for_all_int_types_<double>(-1.3, -0.2, -2);
 }
```

zasdfgbnm replied:
I don't know, it looks like you have declared everything as `C10_HOST_DEVICE` where it should be. And sure, feel free to do it in a separate PR.
